File: android/app/src/main/java/dev/tabscreen/Protocol.kt
package dev.tabscreen

import java.io.DataInputStream
import java.io.EOFException
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Wire protocol shared with the Rust server (see PROTOCOL.md). Little-endian throughout. */
object Protocol {
    const val VERSION = 2

    const val MSG_HELLO = 0x01
    const val MSG_PEN = 0x02
    const val MSG_TOUCH = 0x03
    const val MSG_PING = 0x04
    const val MSG_KEYFRAME_REQUEST = 0x05

    const val MSG_STREAM_CONFIG = 0x81
    const val MSG_VIDEO = 0x82
    const val MSG_AUDIO_CONFIG = 0x83
    const val MSG_PONG = 0x84
    const val MSG_AUDIO = 0x85

    const val MODE_SCREEN = 0
    const val MODE_TOUCHPAD = 1

    const val AUDIO_NONE = 0
    const val AUDIO_TABLET = 1
    const val AUDIO_BOTH = 2

    const val CODEC_H264 = 1
    const val CODEC_HEVC = 2
    const val CODEC_AV1 = 3

    const val VIDEO_FLAG_KEYFRAME = 1

    const val PEN_HOVER_ENTER = 0
    const val PEN_HOVER_MOVE = 1
    const val PEN_HOVER_EXIT = 2
    const val PEN_DOWN = 3
    const val PEN_MOVE = 4
    const val PEN_UP = 5
    const val PEN_CANCEL = 6

    const val TOOL_PEN = 1
    const val TOOL_ERASER = 2

    const val BTN_PRIMARY = 1
    const val BTN_SECONDARY = 2

    const val TOUCH_DOWN = 0
    const val TOUCH_MOVE = 1
    const val TOUCH_UP = 2
    const val TOUCH_CANCEL = 3

    const val PEN_EVENT_SIZE = 36
    const val TOUCH_EVENT_SIZE = 28

    class Frame(val type: Int, val payload: ByteBuffer)

    fun readFrame(input: DataInputStream): Frame {
        val hdr = ByteArray(5)
        input.readFully(hdr)
        val len = ByteBuffer.wrap(hdr, 1, 4).order(ByteOrder.LITTLE_ENDIAN).int
        if (len < 0 || len > 64 shl 20) throw EOFException("bad frame length $len")
        val payload = ByteArray(len)
        input.readFully(payload)
        return Frame(hdr[0].toInt() and 0xff, ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN))
    }

    private fun frame(type: Int, payloadLen: Int): ByteBuffer {
        val b = ByteBuffer.allocate(5 + payloadLen).order(ByteOrder.LITTLE_ENDIAN)
        b.put(type.toByte())
        b.putInt(payloadLen)
        return b
    }

    fun hello(width: Int, height: Int, refresh: Int, codecs: Int, preferred: Int, widthMm: Float, heightMm: Float, mode: Int, audio: Int): ByteArray {
        val b = frame(MSG_HELLO, 2 + 2 + 2 + 2 + 1 + 1 + 4 + 4 + 1 + 1)
        b.putShort(VERSION.toShort())
        b.putShort(width.toShort())
        b.putShort(height.toShort())
        b.putShort(refresh.toShort())
        b.put(codecs.toByte())
        b.put(preferred.toByte())
        b.putFloat(widthMm)
        b.putFloat(heightMm)
        b.put(mode.toByte())
        b.put(audio.toByte())
        return b.array()
    }

    fun keyframeRequest(): ByteArray = frame(MSG_KEYFRAME_REQUEST, 0).array()

    fun ping(tNs: Long): ByteArray {
        val b = frame(MSG_PING, 8)
        b.putLong(tNs)
        return b.array()
    }

    /** Builds a MSG_PEN batch. [count] events must already have been written into [events]. */
    class PenBatch(capacity: Int) {
        private val buf = ByteBuffer.allocate(5 + 2 + capacity * PEN_EVENT_SIZE).order(ByteOrder.LITTLE_ENDIAN)
        private var count = 0

        init { reset() }

        fun reset() {
            buf.clear()
            buf.put(MSG_PEN.toByte()); buf.putInt(0); buf.putShort(0)
            count = 0
        }

        fun add(tNs: Long, action: Int, tool: Int, buttons: Int, x: Float, y: Float, pressure: Float, tiltX: Float, tiltY: Float, distance: Float) {
            if (buf.remaining() < PEN_EVENT_SIZE) return
            buf.putLong(tNs); buf.put(action.toByte()); buf.put(tool.toByte()); buf.put(buttons.toByte()); buf.put(0)
            buf.putFloat(x); buf.putFloat(y); buf.putFloat(pressure); buf.putFloat(tiltX); buf.putFloat(tiltY); buf.putFloat(distance)
            count++
        }

        fun build(): ByteArray? {
            if (count == 0) return null
            buf.putInt(1, buf.position() - 5)
            buf.putShort(5, count.toShort())
            return buf.array().copyOf(buf.position())
        }
    }

    class TouchBatch(capacity: Int) {
        private val buf = ByteBuffer.allocate(5 + 2 + capacity * TOUCH_EVENT_SIZE).order(ByteOrder.LITTLE_ENDIAN)
        private var count = 0

        init { reset() }

        fun reset() {
            buf.clear()
            buf.put(MSG_TOUCH.toByte()); buf.putInt(0); buf.putShort(0)
            count = 0
        }

        fun add(tNs: Long, action: Int, id: Int, x: Float, y: Float, pressure: Float, major: Float) {
            if (buf.remaining() < TOUCH_EVENT_SIZE) return
            buf.putLong(tNs); buf.put(action.toByte()); buf.put(id.toByte()); buf.putShort(0)
            buf.putFloat(x); buf.putFloat(y); buf.putFloat(pressure); buf.putFloat(major)
            count++
        }

        fun build(): ByteArray? {
            if (count == 0) return null
            buf.putInt(1, buf.position() - 5)
            buf.putShort(5, count.toShort())
            return buf.array().copyOf(buf.position())
        }
    }

    fun write(out: OutputStream, msg: ByteArray) {
        out.write(msg)
    }
}

File: server/src/protocol.rs
//! Wire protocol shared with the Android app. All integers little-endian.
//!
//! Frame: `u8 type, u32 payload_len, payload`.
//! See ../../PROTOCOL.md for the human-readable spec; keep both in sync.

pub const VERSION: u16 = 2;

// client -> server
pub const MSG_HELLO: u8 = 0x01;
pub const MSG_PEN: u8 = 0x02;
pub const MSG_TOUCH: u8 = 0x03;
pub const MSG_PING: u8 = 0x04;
pub const MSG_KEYFRAME_REQUEST: u8 = 0x05;
pub const MSG_DISCOVER: u8 = 0x06;

// server -> client
pub const MSG_STREAM_CONFIG: u8 = 0x81;
pub const MSG_VIDEO: u8 = 0x82;
pub const MSG_AUDIO_CONFIG: u8 = 0x83;
pub const MSG_PONG: u8 = 0x84;
pub const MSG_AUDIO: u8 = 0x85;

pub const MODE_SCREEN: u8 = 0;
pub const MODE_TOUCHPAD: u8 = 1;

pub const AUDIO_NONE: u8 = 0;
pub const AUDIO_TABLET: u8 = 1;
pub const AUDIO_BOTH: u8 = 2;

pub const AUDIO_FORMAT_S16LE: u8 = 1;

pub const CODEC_H264: u8 = 1;
pub const CODEC_HEVC: u8 = 2;
pub const CODEC_AV1: u8 = 3;

pub const VIDEO_FLAG_KEYFRAME: u8 = 1;

pub const PEN_HOVER_ENTER: u8 = 0;
pub const PEN_HOVER_MOVE: u8 = 1;
pub const PEN_HOVER_EXIT: u8 = 2;
pub const PEN_DOWN: u8 = 3;
pub const PEN_MOVE: u8 = 4;
pub const PEN_UP: u8 = 5;
pub const PEN_CANCEL: u8 = 6;

pub const TOOL_PEN: u8 = 1;
pub const TOOL_ERASER: u8 = 2;

pub const BTN_PRIMARY: u8 = 1;
pub const BTN_SECONDARY: u8 = 2;

pub const TOUCH_DOWN: u8 = 0;
pub const TOUCH_MOVE: u8 = 1;
pub const TOUCH_UP: u8 = 2;
pub const TOUCH_CANCEL: u8 = 3;


#[derive(Debug, Clone)]
pub struct Hello {
    pub version: u16,
    pub width: u16,
    pub height: u16,
    pub refresh: u16,
    pub codecs: u8,
    pub preferred_codec: u8,
    pub width_mm: f32,
    pub height_mm: f32,
    /// MODE_SCREEN or MODE_TOUCHPAD
    pub mode: u8,
    /// AUDIO_NONE / AUDIO_TABLET / AUDIO_BOTH
    pub audio: u8,
}

#[derive(Debug, Clone, Copy)]
pub struct PenEvent {
    pub t_ns: u64,
    pub action: u8,
    pub tool: u8,
    pub buttons: u8,
    pub x: f32,
    pub y: f32,
    pub pressure: f32,
    pub tilt_x: f32,
    pub tilt_y: f32,
    pub distance: f32,
}

#[derive(Debug, Clone, Copy)]
pub struct TouchEvent {
    pub t_ns: u64,
    pub action: u8,
    pub id: u8,
    pub x: f32,
    pub y: f32,
    pub pressure: f32,
    pub major: f32,
}

pub struct Reader<'a> {
    buf: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    pub fn new(buf: &'a [u8]) -> Self {
        Self { buf, pos: 0 }
    }
    fn take(&mut self, n: usize) -> anyhow::Result<&'a [u8]> {
        if self.pos + n > self.buf.len() {
            anyhow::bail!("truncated message");
        }
        let s = &self.buf[self.pos..self.pos + n];
        self.pos += n;
        Ok(s)
    }
    pub fn u8(&mut self) -> anyhow::Result<u8> {
        Ok(self.take(1)?[0])
    }
    pub fn u16(&mut self) -> anyhow::Result<u16> {
        Ok(u16::from_le_bytes(self.take(2)?.try_into().unwrap()))
    }
    pub fn u64(&mut self) -> anyhow::Result<u64> {
        Ok(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }
    pub fn f32(&mut self) -> anyhow::Result<f32> {
        Ok(f32::from_le_bytes(self.take(4)?.try_into().unwrap()))
    }
}

pub fn parse_hello(p: &[u8]) -> anyhow::Result<Hello> {
    let mut r = Reader::new(p);
    Ok(Hello {
        version: r.u16()?,
        width: r.u16()?,
        height: r.u16()?,
        refresh: r.u16()?,
        codecs: r.u8()?,
        preferred_codec: r.u8()?,
        width_mm: r.f32()?,
        height_mm: r.f32()?,
        mode: r.u8()?,
        audio: r.u8()?,
    })
}

pub fn parse_pen_batch(p: &[u8]) -> anyhow::Result<Vec<PenEvent>> {
    let mut r = Reader::new(p);
    let n = r.u16()? as usize;
    let mut v = Vec::with_capacity(n);
    for _ in 0..n {
        let t_ns = r.u64()?;
        let action = r.u8()?;
        let tool = r.u8()?;
        let buttons = r.u8()?;
        let _pad = r.u8()?;
        v.push(PenEvent {
            t_ns,
            action,
            tool,
            buttons,
            x: r.f32()?,
            y: r.f32()?,
            pressure: r.f32()?,
            tilt_x: r.f32()?,
            tilt_y: r.f32()?,
            distance: r.f32()?,
        });
    }
    Ok(v)
}

pub fn parse_touch_batch(p: &[u8]) -> anyhow::Result<Vec<TouchEvent>> {
    let mut r = Reader::new(p);
    let n = r.u16()? as usize;
    let mut v = Vec::with_capacity(n);
    for _ in 0..n {
        let t_ns = r.u64()?;
        let action = r.u8()?;
        let id = r.u8()?;
        let _pad = r.u16()?;
        v.push(TouchEvent {
            t_ns,
            action,
            id,
            x: r.f32()?,
            y: r.f32()?,
            pressure: r.f32()?,
            major: r.f32()?,
        });
    }
    Ok(v)
}

/// Build a framed message.
pub fn frame(ty: u8, payload: &[u8]) -> Vec<u8> {
    let mut v = Vec::with_capacity(5 + payload.len());
    v.push(ty);
    v.extend_from_slice(&(payload.len() as u32).to_le_bytes());
    v.extend_from_slice(payload);
    v
}

pub fn stream_config(codec: u8, width: u16, height: u16, fps: u16) -> Vec<u8> {
    let mut p = Vec::with_capacity(7);
    p.push(codec);
    p.extend_from_slice(&width.to_le_bytes());
    p.extend_from_slice(&height.to_le_bytes());
    p.extend_from_slice(&fps.to_le_bytes());
    frame(MSG_STREAM_CONFIG, &p)
}

pub fn audio_config(format: u8, rate: u32, channels: u8) -> Vec<u8> {
    let mut p = Vec::with_capacity(6);
    p.push(format);
    p.extend_from_slice(&rate.to_le_bytes());
    p.push(channels);
    frame(MSG_AUDIO_CONFIG, &p)
}

pub fn audio_chunk(pts_ns: u64, data: &[u8]) -> Vec<u8> {
    let mut v = Vec::with_capacity(5 + 8 + data.len());
    v.push(MSG_AUDIO);
    v.extend_from_slice(&((8 + data.len()) as u32).to_le_bytes());
    v.extend_from_slice(&pts_ns.to_le_bytes());
    v.extend_from_slice(data);
    v
}

pub fn video_frame(pts_ns: u64, keyframe: bool, data: &[u8]) -> Vec<u8> {
    let mut v = Vec::with_capacity(5 + 9 + data.len());
    v.push(MSG_VIDEO);
    v.extend_from_slice(&((9 + data.len()) as u32).to_le_bytes());
    v.extend_from_slice(&pts_ns.to_le_bytes());
    v.push(if keyframe { VIDEO_FLAG_KEYFRAME } else { 0 });
    v.extend_from_slice(data);
    v
}
